MINOR: Streams system test fixes/updates

vvcephei opened the pull request:
Some changes required to get the Streams system tests working via Docker

To test: 
```bash
TC_PATHS="tests/kafkatest/tests/streams" bash tests/docker/run_tests.sh
```
That command will take about 3.5 hours, and should pass. Note there are a couple of ignored tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
MINOR: Streams system tests fixes/updates

* add 1.0.0 and 1.0.1

Closes #4689

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -163,7 +163,7 @@ def __init__(self, test_context, kafka, command):
     def clean_node(self, node):
         if self.clean_node_enabled:
-            super.clean_node(self, node)
+            super(StreamsEosTestBaseService, self).clean_node(node)
```
So, this wasn't valid python3... It's also the only `super` invocation in kafkatest that's formatted this way. I'm not sure if it's valid python2.

But it only affects cleanup for the Eos test, and it just generates a warning instead of failing the test. So I can see how this could go unnoticed if the tests are usually run from a clean state instead of cleaning up after themselves.

vvcephei commented on tests/setup.py:
```diff
@@ -38,7 +38,7 @@ def finalize_options(self):
         # import here, cause outside the eggs aren't loaded
         import pytest
-        print self.pytest_args
+        print(self.pytest_args)
```
This was not valid python3. Maybe everyone else ran setup.py in python2 and haven't run it again since?

vvcephei replied:
note that it's also valid python2 to use parens here, so I'd advocate for keeping this change.

vvcephei commented on .gitignore:
```diff
@@ -47,3 +47,7 @@ tests/venv
 .release-settings.json
+
+kafkatest.egg-info/
+systest/
```
setup.py generates these directories. I assume we don't want them checked in.

vvcephei commented on tests/docker/Dockerfile:
```diff
@@ -40,12 +40,17 @@ RUN ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && cp -f /root/.ssh/id_rsa.p
+# Streams test dependencies
+RUN curl -s "$KAFKA_MIRROR/kafka-streams-0.10.1.1-test.jar" -o /opt/kafka-0.10.1.1/libs/kafka-streams-0.10.1.1-test.jar && \
+    curl -s "$KAFKA_MIRROR/kafka-streams-0.10.2.1-test.jar" -o /opt/kafka-0.10.2.1/libs/kafka-streams-0.10.2.1-test.jar && \
+    curl -s "$KAFKA_MIRROR/kafka-streams-0.11.0.0-test.jar" -o /opt/kafka-0.11.0.0/libs/kafka-streams-0.11.0.0-test.jar
```
These test deps were missing. I guess no one has had to run the upgrade/downgrade test in docker. (Jenkins still uses Vagrant)

mjsax commented:
@vvcephei Can you trigger the branch builder to see if this changes work there? Thx.

guozhangwang commented on tests/docker/Dockerfile:
```diff
@@ -40,12 +40,17 @@ RUN ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && cp -f /root/.ssh/id_rsa.p
+RUN mkdir -p "/opt/kafka-0.8.2.2" && chmod a+rw /opt/kafka-0.8.2.2 && curl -s "$KAFKA_MIRROR/kafka_2.10-0.8.2.2.tgz" | tar xz --strip-components=1 -C "/opt/kafka-0.8.2.2"
+RUN mkdir -p "/opt/kafka-0.9.0.1" && chmod a+rw /opt/kafka-0.9.0.1 && curl -s "$KAFKA_MIRROR/kafka_2.11-0.9.0.1.tgz" | tar xz --strip-components=1 -C "/opt/kafka-0.9.0.1"
+RUN mkdir -p "/opt/kafka-0.10.0.1" && chmod a+rw /opt/kafka-0.10.0.1 && curl -s "$KAFKA_MIRROR/kafka_2.11-0.10.0.1.tgz" | tar xz --strip-components=1 -C "/opt/kafka-0.10.0.1"
+RUN mkdir -p "/opt/kafka-0.10.1.1" && chmod a+rw /opt/kafka-0.10.1.1 && curl -s "$KAFKA_MIRROR/kafka_2.11-0.10.1.1.tgz" | tar xz --strip-components=1 -C "/opt/kafka-0.10.1.1"
```
I think the `chmod` is done in `vagrant/base.sh`?

vvcephei replied:
yeah, it is. That's what I found that ultimately tipped me off to what was wrong with the docker code. (it starts at line 104 there)

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -123,6 +123,7 @@ def start_cmd(self, node):
               "INCLUDE_TEST_JARS=true %(kafka_run_class)s %(streams_class_name)s " \
               " %(kafka)s %(state_dir)s %(user_test_args)s %(user_test_args1)s %(user_test_args2)s" \
               " %(user_test_args3)s & echo $! >&3 ) 1>> %(stdout)s 2>> %(stderr)s 3> %(pidfile)s" % args
+        print(cmd)
```
Having this in the output reduced my debug cycle time a lot. I've stripped out all my other debugging output. What do you think about leaving this in?

vvcephei commented on tests/kafkatest/services/streams.py:
```diff
@@ -123,6 +123,7 @@ def start_cmd(self, node):
               "INCLUDE_TEST_JARS=true %(kafka_run_class)s %(streams_class_name)s " \
               " %(kafka)s %(state_dir)s %(user_test_args)s %(user_test_args1)s %(user_test_args2)s" \
               " %(user_test_args3)s & echo $! >&3 ) 1>> %(stdout)s 2>> %(stderr)s 3> %(pidfile)s" % args
+        self.logger.info("Executing: " + cmd)
```
I got a lot of value out of having this in the output. What do you think about leaving this in?

bbejeck replied:
+1 from me

guozhangwang replied:
+1.

vvcephei commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -15,13 +15,14 @@
+from kafkatest.version import LATEST_0_10_1, LATEST_0_10_2, LATEST_0_11, LATEST_1_0, DEV_BRANCH, KafkaVersion
 import time
 
+upgrade_versions = [str(LATEST_0_10_2), str(LATEST_0_11), str(LATEST_1_0), str(DEV_BRANCH)]
```
Note: I had to drop 0.10.1 because `StreamsSmokeTest` had a different package name in that version (was: `org.apache.kafka.streams.smoketest.StreamsSmokeTest`.

I can put it back in, but it would require a conditional in the streams.py::StreamsSmokeTestBaseService. Or we can leave it out.

Thoughts?

bbejeck replied:
I would say leave it out, I believe the change from `0.10.1` to a higher version requires a full stop and can't be done from a rolling restart.

guozhangwang replied:
Actually from `0.10.0` to `0.10.1` requires a full stop, from `0.10.1` to newer version does not. But I think it is still fine to leave it out.

vvcephei commented:
The most substantial change here is the switch from `@parameterize` to `@matrix` to ensure we cover every combination of versions.

Downside: this increases the test run time by at least 1 hour!

We can:
* accept that
* prune the matrix (select just a handful of scenarios like before)
* fork multiple jenkins builds of different parts of that matrix to run them in parallel

Thoughts?

vvcephei commented:
@guozhangwang @bbejeck @mjsax Please take another look at this. It passes on my machine (using docker). I'm running an all-streams jenkins test to make sure it still passes using vagrant and also to measure the new run time.

bbejeck commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -94,74 +89,79 @@ def test_upgrade_downgrade_streams(self, from_version, to_version):
+        if from_version != to_version:
+            # Setup phase
+            self.zk = ZookeeperService(self.test_context, num_nodes=1)
+            self.zk.start()
```
I think you can do the zookeeper start in the `setUp` method

vvcephei replied:
hm. actually, there isn't a `setUp` method... Are you suggesting I make one?

FWIW, I think it's actually fine like this. The test is linearly readable this way.

bbejeck replied:
There's a `setUp` method on the [Test](https://github.com/confluentinc/ducktape/blob/master/ducktape/tests/test.py#L98) class that we can override.  It's fine like it is, but most of the other tests do ZK start-up in the `setUp` method, so I was aiming for consistency.

vvcephei replied:
aha, I gotcha. sure, I can do that.

guozhangwang replied:
+1

vvcephei replied:
Ok, I'm glad that I know what you were talking about now. I actually think in this case, it's a little better as is, since we want to start the broker with different versions during the test.

We could still do it, but then we'd have to stop the broker and upgrade it and restart it during the test, or only start zk in startUp and start the broker during the tests... I think in this case, it's a more complicated control flow in the tests in exchange for diminishing returns.

I can totally get it for tests that just need zk and kafka running and don't care about the version, but in this case, I think I like it better flat.

bbejeck commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -94,74 +89,79 @@ def test_upgrade_downgrade_streams(self, from_version, to_version):
+        if from_version != to_version:
+            # Setup phase
+            self.zk = ZookeeperService(self.test_context, num_nodes=1)
+            self.zk.start()
```
same as above

guozhangwang commented on tests/kafkatest/tests/streams/streams_upgrade_test.py:
```diff
@@ -15,13 +15,14 @@
 from kafkatest.services.zookeeper import ZookeeperService
 from kafkatest.services.streams import StreamsSmokeTestDriverService, StreamsSmokeTestJobRunnerService
-from kafkatest.version import LATEST_0_10_1, LATEST_0_10_2, LATEST_0_11_0, DEV_BRANCH, KafkaVersion
+from kafkatest.version import LATEST_0_10_1, LATEST_0_10_2, LATEST_0_11, LATEST_1_0, DEV_BRANCH, KafkaVersion
```
Remove `LATEST_0_10_1` as you mentioned below?

guozhangwang reviewed (commented):
Just a note that @bbejeck is working on https://github.com/apache/kafka/pull/4690/files which may have some overlap on the class files, some we would need some rebasing either one way or the other.

vvcephei commented:
@guozhangwang Thanks. Yeah, I'm prepared to rebase on #4690 once it's merged.

I'm running a full streams system test again now, just to be double-sure that it passes before recommending a merge.

vvcephei commented:
I've just added 1.0.1, since it's in the mirror now.
I also rebased onto trunk.
I'm running the sys tests once more to be sure they work with 1.0.1, but otherwise, this will be ready to merge.

vvcephei commented:
Alright, @mjsax @guozhangwang, I think this thing is ready to merge. I re-ran the streams system tests (both locally:docker and in jenkins:vagrant), and they actually all passed this time!

guozhangwang reviewed (approved):
LGTM!
